# Лабораторная работа 3.1

## Счастливый билетик

![](http://f2.mylove.ru/UNiHT7dRJ4.jpg)

Билет считается счастливым, если выполнено следующее условие: сумма первых трёх цифр номера равна сумме последних трёх цифр.

**Задание:**

1) Определите число `ticket_number` — шестизначный номер билета;

2) Напишите код, который по шестизначному номеру `ticket_number` билетика проверяет, является ли он счастливым;

3) Если номер счастливый, выведите строку `Yes`, иначе — `No`.

---

**Пример 1:**

*Input:* 123456

*Output:* No


**Пример 2:**

*Input:* 123042

*Output:* Yes

In [1]:
ticket_number = [int(i) for i in input("Введите номер билета: ")] 

Введите номер билета: 123042


In [2]:
if sum(ticket_number[:3]) == sum(ticket_number[3:]):
    print("yes")
else:
    print("no")

yes


## Пароль

![](http://forum.ixbt.com/post.cgi?id=attach:77:14138:1341:1)

Пусть пароль может содержать только латинские буквы, знаки препинания и цифры.

Пароль считается надёжным, если удовлетворяет следующим условиям:
- содержит буквы в разных регистрах;
- содержит цифры;
- содержит не менее `4` уникальных символов;
- не содержит ваше имя латинницей, записанное буквами любых регистров (anna, iVan, ...).

Иначе пароль считается слабым.

**Задание:**

1) Определите строку `password` — придуманный вами пароль;

2) Напишите код, который по паролю `password` проверяет, является ли он надёжным;

3) Если пароль надёжный, выведите строку `strong`, иначе — `weak`.


---

Пусть имя пользоватея -- Андрей.

**Пример 1:**

*Input:* Aandrei123

*Output:* weak


**Пример 2:**

*Input:* an12dRei

*Output:* strong

In [3]:
name = input("Введите ваше имя: ")
password = input("Введите пароль: ")

Введите ваше имя: Dmitry
Введите пароль: Dmi123try


In [4]:
unic = set(password)
if password.isalpha() or password.isdigit():
    print("weak")
    exit(-1)
if password.isupper() or password.islower():
    print("weak")
    exit(-1)

if name.lower() in password.lower():
    print("weak")
    exit(-1)
if len(unic) < 4:
    print("weak")
    exit(-1)  
else:
    print("strong")

strong


## Числа Фибоначчи
Как известно, [числа Фибоначчи](https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%B0_%D0%A4%D0%B8%D0%B1%D0%BE%D0%BD%D0%B0%D1%87%D1%87%D0%B8) — это последовательность чисел, каждое из которых равно сумме двух предыдущих (первые два числа равны $1$):
$$
    1, 1, 2, 3, 5, 8, 13, \ldots
$$

**Задание:**

1) Определите число `amount` — количество чисел Фибоначчи, которые надо вывести;

2) Напишите код, который выводит первые `amount` чисел Фибоначчи.

---

**Пример 1:**

*Input:* 3

*Output:* 1 1 2


**Пример 2:**

*Input:* 10

*Output:* 1 1 2 3 5 8 13 21 34 55


In [7]:
amount = int(input("Введите количество чисел: "))

Введите количество чисел: 15


In [8]:
import math


def fibonacci(amount):
    phi = (1 + math.sqrt(5)) / 2
    fib_list = [(phi**n - (1-phi)**n) / math.sqrt(5) for n in range(amount)]
    return fib_list


fib_list = fibonacci(amount)
fib_str = ' '.join(str(int(fib)) for fib in fib_list)
print(fib_str)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377


## Время исследований

![](https://2.bp.blogspot.com/-VOT5x12UqPY/WSh53tQ94HI/AAAAAAAAAMY/-nlcz28_Fj4UzJjbn4bBINKCnzwag1W4QCLcB/s1600/my-pic-widget.jpg)

На сайте https://www.kaggle.com/ выберите любой набор данных в формате CSV и проведите для него маленькое исследование: загрузите данные из набора с использованием стандартного модуля `csv`, посмотрите средние значения и стандартные отклонения двух выбранных числовых атрибутов, найдите [методом наименьших квадратов](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D1%82%D0%BE%D0%B4_%D0%BD%D0%B0%D0%B8%D0%BC%D0%B5%D0%BD%D1%8C%D1%88%D0%B8%D1%85_%D0%BA%D0%B2%D0%B0%D0%B4%D1%80%D0%B0%D1%82%D0%BE%D0%B2) уравнение линейной зависимости, связывающей один числовой атрибут с другим. Для оценки заданной зависимости найдите [коэффициент парной корреляции](https://ru.wikipedia.org/wiki/%D0%9A%D0%BE%D1%80%D1%80%D0%B5%D0%BB%D1%8F%D1%86%D0%B8%D1%8F), сделайте соответствующие выводы.

**Результаты надо обязательно прокомментировать и пояснить!**

---

**Пример 1:**

Пусть таблица `bikes.csv` содержит данные по арендам велосипедов за $2$ года:
- `datetime`: дата и время аренды
- `season`: время года
- `temp`: температура воздуха по Цельсию
- `windspeed`: скорость ветра 
- `registered`: число аренд

Одно из направлений исследования могло бы заключаться в проверке зависимости суммарного числа аренд от температуры воздуха.

*в такой ячейке (режим Markdown) можно писать текст*

In [9]:
import csv
from math import sqrt


with open('dth_to_pol.csv', 'r', newline='') as csvfile:
    data = csv.reader(csvfile, delimiter=',')
    afg_total_pol = []
    afg_ozone_pol = []
    for row in data:
        if row[0] == "Afghanistan":
            afg_total_pol.append(float(row[3]))
            afg_ozone_pol.append(float(row[6]))

Затем нам необходимо посчитать среднее значение коэффициентов из списка

In [10]:
mean_total = sum(afg_total_pol) / len(afg_total_pol)
mean_ozone = sum(afg_ozone_pol) / len(afg_ozone_pol)
print("Среднее значение коэффициента загрязнения воздуха в Афганистане с "
      f"1990 по 2017 год: {mean_total}"
)
print("Среднее значение коэффициента загрязнения воздуха озоном в "
      "Афганистане с "
      f"1990 по 2017 год: {mean_ozone}"
)

Среднее значение коэффициента загрязнения воздуха в Афганистане с 1990 по 2017 год: 252.84255072458163
Среднее значение коэффициента загрязнения воздуха озоном в Афганистане с 1990 по 2017 год: 5.797505605371072


Для того, чтобы посчитать стандартное отклонение выбранных величин создадим две промежуточные переменные, которые будут содержать в себе сумму квадратов разности для каждого элемента из списка, поделённую на количество элементов.

In [11]:
var1 = sum((elem-mean_total)**2 for elem in afg_total_pol) / len(afg_total_pol)
st_dev_total = sqrt(var1)
var2 = sum((elem-mean_ozone)**2 for elem in afg_ozone_pol) / len(afg_ozone_pol)
st_dev_ozone = sqrt(var2)
print("Стандартное отклонение коэффициента общего загрязнения воздуха: "
      f"{st_dev_total}"
)
print("Стандартное отклонение коэффициента загрязнения воздуха озоном: " 
      f"{st_dev_ozone}"
)

Стандартное отклонение коэффициента общего загрязнения воздуха: 37.35656399089619
Стандартное отклонение коэффициента загрязнения воздуха озоном: 0.09717587256065482


Для того, чтобы найти коэффициенты уравнения линейной зависимости посчитает суммму произведений элементов из двух списков, а также сумму квадратов элементов из списка коэффициентов общего загрязнения воздуха. Затем посчитает по формуле коэффициент k и b. После выводим их на экран, а также значения функции для каждой точки спсика коэффициентов общего загрязнения воздуха

In [12]:
sum_ab = 0
sum_square = 0
for idx, elem in enumerate(afg_total_pol):
    sum_ab += elem * afg_ozone_pol[idx]
    sum_square += elem**2
size = len(afg_total_pol)
k_lin = (size * sum_ab - sum(afg_total_pol) * sum(afg_ozone_pol))/(size * sum_square - sum(afg_total_pol)**2)
b_lin = mean_ozone - mean_total * k_lin
func_val = []
for elem in afg_total_pol:
    func_val.append(k_lin * elem + b_lin)
print(f"Уравнение линейной зависимости: y = {k_lin}x + {b_lin}")
print("Значения функции на кривой методом наименьших квадратов: ")
for val in func_val:
    print(val)

Уравнение линейной зависимости: y = -0.00048149182135542336x + 5.919247225635601
Значения функции на кривой методом наименьших квадратов: 
5.77505135072677
5.778999266912121
5.7849287958966285
5.785011728466223
5.780980626729237
5.780570732469591
5.78123116360491
5.781325119009701
5.781412872176521
5.781234327095285
5.782606457244574
5.783522615764744
5.787637650720681
5.788617094747192
5.789676076564802
5.793801344799917
5.7963822528383595
5.800549250447244
5.804888299644632
5.809102509246886
5.812584474584094
5.816047786447732
5.819391261674628
5.822512803044716
5.825145928668351
5.826896555523149
5.829367653181076
5.830680952160238


Для рассчёта коэффициента парной корреляции разделим формулу на числитель и знаменатель. Для числителя посчитаем произведение разностей элементов каждого из списков со средним значением этих списков. В знаменателе считаем произведение средних квадратических отклонений.

In [13]:
cor_chisl = 0
for idx, elem in enumerate(afg_total_pol):
    cor_chisl += (elem - mean_total)*(afg_ozone_pol[idx] - mean_ozone)
sqr_diff_total = sum((elem-mean_total)**2 for elem in afg_total_pol)
sqr_diff_ozone = sum((elem-mean_ozone)**2 for elem in afg_ozone_pol)
r_xy = cor_chisl / sqrt(sqr_diff_total * sqr_diff_ozone)
print(f"Коэффициент парной корреляции: {r_xy}")

Коэффициент парной корреляции: -0.185096151561045


Значение -0.185 можно рассматривать как почти полное отсутствие корреляции.